a completed flux that's 'hot' via RefCountGrace fails when using next

```
        VirtualTimeScheduler scheduler = VirtualTimeScheduler.create();
        Duration gracePeriod = Duration.ofMillis(10);
        Flux<String> f = Flux.just("hello world")
                             .replay(1)
                             .refCount(1, gracePeriod, scheduler);

        f.subscribe();

        StepVerifier.create(f.next())
                    .expectNext("hello world")
                    .verifyComplete();

        scheduler.advanceTimeBy(gracePeriod);
```

This throws

```
reactor.core.Exceptions$BubblingException: java.util.concurrent.CancellationException: Disconnected
	at reactor.core.Exceptions.bubble(Exceptions.java:112)
	at reactor.core.publisher.Operators.onErrorDropped(Operators.java:263)
	at reactor.core.publisher.LambdaSubscriber.onError(LambdaSubscriber.java:116)
	at reactor.core.publisher.FluxRefCountGrace$RefCountInner.onError(FluxRefCountGrace.java:203)
	at reactor.core.publisher.FluxReplay$ReplaySubscriber.disconnectAction(FluxReplay.java:1180)
	at reactor.core.publisher.FluxReplay$ReplaySubscriber.dispose(FluxReplay.java:1172)
	at reactor.core.publisher.Disposables.dispose(Disposables.java:147)
	at reactor.core.publisher.FluxRefCountGrace.timeout(FluxRefCountGrace.java:137)
	at reactor.core.publisher.FluxRefCountGrace$RefConnection.run(FluxRefCountGrace.java:165)
	at reactor.test.scheduler.VirtualTimeScheduler.advanceTime(VirtualTimeScheduler.java:297)
	at reactor.test.scheduler.VirtualTimeScheduler.advanceTimeBy(VirtualTimeScheduler.java:205)
Caused by: java.util.concurrent.CancellationException: Disconnected
	at reactor.core.publisher.FluxReplay$ReplaySubscriber.disconnectAction(FluxReplay.java:1177)
	... 29 more
```

that felt like a "safe" sequence of operators

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java
@@ -119,7 +119,14 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 				c = u;
 			}
 
-			if (c.trySubscribe(inner)) {
+			if (c.add(inner)) {
+				if (inner.isCancelled()) {
+					c.remove(inner);
+				}
+				else {
+					inner.parent = c;
+				}
+				c.drain();
 				break;
 			}
 		}
@@ -360,20 +367,6 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 
 		boolean tryConnect() {
 			return connected == 0 && CONNECTED.compareAndSet(this, 0, 1);
-		}
-
-		boolean trySubscribe(PublishInner<T> inner) {
-			if (add(inner)) {
-				if (inner.isCancelled()) {
-					remove(inner);
-				}
-				else {
-					inner.parent = this;
-				}
-				drain();
-				return true;
-			}
-			return false;
 		}
 
 		final void drain() {

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxReplay.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxReplay.java
@@ -1011,9 +1011,9 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 		ReplaySubscriber<T> s;
 		for (; ; ) {
 			s = connection;
-			if (s == null || s.isTerminated()) {
+			if (s == null) {
 				ReplaySubscriber<T> u = newState();
-				if (!CONNECTION.compareAndSet(this, s, u)) {
+				if (!CONNECTION.compareAndSet(this, null, u)) {
 					continue;
 				}
 
@@ -1033,25 +1033,29 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 	@Override
 	public void subscribe(Subscriber<? super T> s) {
 		ReplayInner<T> inner = new ReplayInner<>(s);
-		s.onSubscribe(inner);
 		for (; ; ) {
-			if (inner.isCancelled()) {
-				break;
-			}
-
 			ReplaySubscriber<T> c = connection;
-			if (c == null || c.isTerminated()) {
+			if (c == null) {
 				ReplaySubscriber<T> u = newState();
-				if (!CONNECTION.compareAndSet(this, c, u)) {
+				if (!CONNECTION.compareAndSet(this, null, u)) {
 					continue;
 				}
 
 				c = u;
 			}
 
-			if (c.trySubscribe(inner)) {
-				break;
+			c.add(inner);
+
+			s.onSubscribe(inner);
+
+			if (inner.isCancelled()) {
+				c.remove(inner);
 			}
+
+			inner.parent = c;
+			c.buffer.replay(inner);
+
+			break;
 		}
 	}
 
@@ -1137,9 +1141,7 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 			else {
 				b.onError(t);
 
-				ReplaySubscription<T>[] a = subscribers;
-
-				for (ReplaySubscription<T> rs : a) {
+				for (ReplaySubscription<T> rs : terminate()) {
 					b.replay(rs);
 				}
 			}
@@ -1151,9 +1153,7 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 			if (!b.isDone()) {
 				b.onComplete();
 
-				ReplaySubscription<T>[] a = subscribers;
-
-				for (ReplaySubscription<T> rs : a) {
+				for (ReplaySubscription<T> rs : terminate()) {
 					b.replay(rs);
 				}
 			}
@@ -1166,18 +1166,16 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 			}
 			if (Operators.terminate(S, this)) {
 				cancelled = true;
-				if (WIP.getAndIncrement(this) != 0) {
-					return;
-				}
-				disconnectAction();
-			}
-		}
 
-		void disconnectAction() {
-			CancellationException ex = new CancellationException("Disconnected");
-			buffer.onError(ex);
-			for (ReplaySubscription<T> inner : terminate()) {
-				inner.actual().onError(ex);
+				CONNECTION.lazySet(parent, null);
+
+				CancellationException ex = new CancellationException("Disconnected");
+				buffer.onError(ex);
+
+				for (ReplaySubscription<T> inner : terminate()) {
+					buffer.replay(inner);
+				}
+
 			}
 		}
 
@@ -1263,20 +1261,6 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 			return connected == 0 && CONNECTED.compareAndSet(this, 0, 1);
 		}
 
-		boolean trySubscribe(ReplayInner<T> inner) {
-			if (add(inner)) {
-				if (inner.isCancelled()) {
-					remove(inner);
-				}
-				else {
-					inner.parent = this;
-					buffer.replay(inner);
-				}
-				return true;
-			}
-			return false;
-		}
-
 		@Override
 		@Nullable
 		public Object scanUnsafe(Attr key) {
@@ -1298,7 +1282,7 @@ final class FluxReplay<T> extends ConnectableFlux<T> implements Scannable, Fusea
 
 		@Override
 		public boolean isDisposed() {
-			return cancelled || buffer.isDone();
+			return cancelled;
 		}
 
 	}

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxRefCountGraceTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxRefCountGraceTest.java
@@ -25,6 +25,7 @@ import reactor.core.Scannable;
 import reactor.core.scheduler.Schedulers;
 import reactor.test.StepVerifier;
 import reactor.test.publisher.TestPublisher;
+import reactor.test.scheduler.VirtualTimeScheduler;
 import reactor.test.subscriber.AssertSubscriber;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -204,6 +205,28 @@ public class FluxRefCountGraceTest {
 		//since the TestPublisher doesn't cleanup on termination, we can re-evaluate the doFinally
 		publisher.complete();
 		assertThat(termination.get()).isEqualTo(SignalType.ON_COMPLETE);
+	}
+
+	@Test
+	public void shouldNotRetainSubscriptionToSourceWhenComplete() throws Exception {
+		VirtualTimeScheduler scheduler = VirtualTimeScheduler.create();
+		Duration gracePeriod = Duration.ofMillis(10);
+		Flux<String> f = Flux.just("hello world")
+		                     .replay(1)
+		                     .refCount(1, gracePeriod, scheduler);
+
+		AssertSubscriber<String> s = AssertSubscriber.create();
+		f.subscribe(s);
+
+		scheduler.advanceTimeBy(gracePeriod);
+
+		StepVerifier.create(f.next())
+		            .expectNext("hello world")
+		            .verifyComplete();
+
+		scheduler.advanceTimeBy(gracePeriod);
+
+		s.assertValueCount(1).assertNoError().assertComplete();
 	}
 
 	@Test

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxReplayTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxReplayTest.java
@@ -28,6 +28,7 @@ import org.reactivestreams.Subscriber;
 import org.reactivestreams.Subscription;
 
 import reactor.core.Disposable;
+import reactor.core.Exceptions;
 import reactor.core.Fuseable;
 import reactor.core.Scannable;
 import reactor.core.scheduler.Schedulers;
@@ -247,7 +248,8 @@ public class FluxReplayTest extends FluxOperatorTest<String, String> {
 		                                         .delayElements(Duration.ofMillis(1000))
 		                                         .replay(2, Duration.ofMillis(2000))
 		                                         .autoConnect()
-		                                         .elapsed();
+		                                         .elapsed()
+				.log();
 
 		StepVerifier.create(source)
 		            .expectFusion(Fuseable.ANY)
@@ -267,10 +269,10 @@ public class FluxReplayTest extends FluxOperatorTest<String, String> {
 
 	@Test
 	public void cancel() {
-		ConnectableFlux<Integer> replay = Flux.range(1, 5)
+		ConnectableFlux<Integer> replay = UnicastProcessor.<Integer>create()
 		                                             .replay(2);
-		Disposable subscribed = replay.subscribe(v -> {},
-				e -> fail(e.toString()));
+
+		replay.subscribe(v -> {}, e -> { throw Exceptions.propagate(e); });
 
 		Disposable connected = replay.connect();
 
@@ -278,7 +280,7 @@ public class FluxReplayTest extends FluxOperatorTest<String, String> {
 		//propagated by connect().dipose()
 		assertThatExceptionOfType(RuntimeException.class)
 				.isThrownBy(connected::dispose)
-	            .withMessage("java.util.concurrent.CancellationException: Disconnected");
+	            .withMessage("Disconnected");
 
 		boolean cancelled = ((FluxReplay.ReplaySubscriber) connected).cancelled;
 		assertThat(cancelled).isTrue();
@@ -299,7 +301,9 @@ public class FluxReplayTest extends FluxOperatorTest<String, String> {
         FluxReplay<Integer> main = new FluxReplay<>(Flux.just(1), 2, 1000, Schedulers.single());
         FluxReplay.ReplayInner<Integer> test = new FluxReplay.ReplayInner<>(actual);
         FluxReplay.ReplaySubscriber<Integer> parent = new FluxReplay.ReplaySubscriber<>(new FluxReplay.UnboundedReplayBuffer<>(10), main);
-        parent.trySubscribe(test);
+        parent.add(test);
+        test.parent = parent;
+        parent.buffer.replay(test);
 
         Assertions.assertThat(test.scan(Scannable.ScannableAttr.PARENT)).isSameAs(parent);
         Assertions.assertThat(test.scan(Scannable.ScannableAttr.ACTUAL)).isSameAs(actual);

